src/globalvars.cpp

```diff
@@ -76,18 +126,151 @@ void update_global_vars(const struct bpf_object *bpf_object,
+        "Failed to set size from " + std::to_string(actual_size) + " to " +
+        std::to_string(desired_size) + " for section " + section_name);
+  }
+
+  // No need to memset to zero as we memset on each usage
+
+  // Verify we can still find variable name via BTF and it hasn't been cleared
+  // after size changes
+  auto vars_and_offset = find_btf_var_offsets(bpf_object,
+                                              section_name,
+                                              needed_global_vars);
+  if (vars_and_offset.at(global_var) != 0) {
+    LOG(BUG) << "Read-write global variable " << to_string(global_var)
+             << " must be at offset 0 in section " << section_name;
+  }
+}
+
+void update_global_vars(const struct bpf_object *bpf_object,
+                        const std::unordered_map<std::string, struct bpf_map *>
+                            &section_name_to_global_vars_map,
+                        const BPFtrace &bpftrace)
+{
+  verify_maps_found(section_name_to_global_vars_map, bpftrace);
+  for (const auto &[section_name, global_vars_map] :
+       section_name_to_global_vars_map) {
+    const auto needed_global_variables = get_global_vars_for_section(
+        section_name, bpftrace);
+    if (needed_global_variables.empty()) {
+      continue;
+    }
+    if (section_name == RO_SECTION_NAME) {
+      update_global_vars_rodata(bpf_object,
+                                section_name,
+                                global_vars_map,
+                                needed_global_variables,
+                                bpftrace);
+    } else {
+      update_global_vars_custom_rw_section(bpf_object,
+                                           section_name,
+                                           global_vars_map,
+                                           needed_global_variables,
+                                           bpftrace);
     }
   }
 }
 
+std::string to_string(GlobalVar global_var)
+{
+  return get_config(global_var).name;
+}
+
+GlobalVar from_string(std::string_view name)
+{
+  for (const auto &[global_var, config] : GLOBAL_VAR_CONFIGS) {
+    if (config.name == name)
+      return global_var;
+  }
+  LOG(BUG) << "Unknown global variable " << name;
+  return {}; // unreachable
+}
+
+static SizedType make_rw_type(size_t num_elements,
+                              const SizedType &element_type)
+{
+  auto subtype = CreateArray(num_elements, element_type);
+  // For 1 CPU, will be adjusted to actual CPU count at runtime
+  return CreateArray(1, subtype);
+}
+
+SizedType get_type(bpftrace::globalvars::GlobalVar global_var,
+                   const RequiredResources &resources)
+{
+  switch (global_var) {
+    case bpftrace::globalvars::GlobalVar::NUM_CPUS:
+    case bpftrace::globalvars::GlobalVar::MAX_CPU_ID:
+      return CreateInt64();
+    case bpftrace::globalvars::GlobalVar::FMT_STRINGS_BUFFER:
+      assert(resources.max_fmtstring_args_size > 0);
+      return make_rw_type(
```

Comment:
Question: so in future PRs, there will be other things calling `make_rw_type` with more than 1 num_elements?

Reply:
Yup!

2 use cases:
1. String scratch map already in code: https://github.com/bpftrace/bpftrace/blob/f6d6345934d92be18de3fc79a16d5163f9d93f6a/src/ast/passes/codegen_llvm.cpp#L3782
2. Tuple scratch map from PR (aka size = max tuple depth): https://github.com/bpftrace/bpftrace/pull/3456